File: server/src/llm/providers/llm-provider.factory.ts
import { Injectable } from '@nestjs/common';
import { LlmInterfaces } from 'src/types/reports.types';
import { ILlmProviderHandler } from './llm-provider.interface';
import { OpenAiHandler } from './handlers/openai.handler';
import { OllamaHandler } from './handlers/ollama.handler';
import { ModuleRef } from '@nestjs/core';

@Injectable()
export class LlmProviderFactory {
  constructor(private readonly moduleRef: ModuleRef) {}

  create(llmInterface: LlmInterfaces): ILlmProviderHandler {
    switch (llmInterface) {
      case LlmInterfaces.OpenAi: {
        return this.moduleRef.get(OpenAiHandler);
      }

      case LlmInterfaces.Ollama: {
        return this.moduleRef.get(OllamaHandler);
      }
    }
  }
}

File: src/ReportsCheck.Infrastructure/Llm/LlmProviderFactory.cs
using Microsoft.Extensions.DependencyInjection;
using ReportsCheck.Infrastructure.Llm.Handlers;
using LlmInterfaceEnum = ReportsCheck.Domain.Enums.LlmInterface;

namespace ReportsCheck.Infrastructure.Llm;

/// <summary>
/// Выбор обработчика по интерфейсу модели. Порт LlmProviderFactory.
/// </summary>
public class LlmProviderFactory : ILlmProviderFactory
{
    private readonly IServiceProvider _serviceProvider;

    public LlmProviderFactory(IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
    }

    public ILlmProviderHandler Create(LlmInterfaceEnum llmInterface) => llmInterface switch
    {
        LlmInterfaceEnum.OpenAi => _serviceProvider.GetRequiredService<OpenAiHandler>(),
        LlmInterfaceEnum.Ollama => _serviceProvider.GetRequiredService<OllamaHandler>(),
        _ => throw new ArgumentOutOfRangeException(nameof(llmInterface), llmInterface, "Неизвестный LLM-интерфейс"),
    };
}
